R/cube-result.R

```diff
@@ -13,9 +13,248 @@ setMethod("initialize", "CrunchCube", function (.Object, ...) {
+                    length(out[[i - 1]]) == 1 && # MR response variable is a single number
+                    !isTRUE(out[[i - 1]])) {
+                if (x@useNA == "no") {
+                    # This is used by skipMissingCategories below
+                    out[i] <- "mr_select_drop"
+                } else {
+                    # assign index value to "Selected" along the mr_selection dimension
+                    out[i] <- 1
+                }
+            }
+        }
+    }
+    return(out)
+}
+
+
+#' Handle missing categories in CrunchCube
+#'
+#' By default we don't display missing categories. The result is that when the
+#' user subsets a cube with a missing category, we need to translate that subset
+#' to only refer to the non-missing categories. This only occurs if
+#' `cube@useMissing` is set to `"no"`. This function handles this behavior by
+#' translating the user supplied indices to logical vectors.
+#'
+#' @param cube a CrunchCube
+#' @param index a index of the real cube which was generated by
+#'   `translateCubeIndex()`
+#' @return A list of logical vectors
+#' @keywords internal
+skipMissingCategories <- function(cube, index){
+    visible_cats <- evalUseNA(cube@arrays$count, dims = cube@dims, useNA = cube@useNA)
+    mapply(function(visable, sub){
+        if (identical(sub, "mr_select_drop")) {
+            # select the "Selected" element of the selection dimension.
+            return(c(TRUE, FALSE, FALSE))
+        }
+        if (isTRUE(sub)) {
+            out <- rep(TRUE, length(visable))
+        } else {
+            out <- rep(FALSE, length(visable))
+            out[visable][sub] <- rep(TRUE, length(sub))
+        }
+        return(out)}, visable = visible_cats, sub = index, SIMPLIFY = FALSE)
+}
+
+subsetByList <- function(arr, arglist, drop){
+    #suppresses "named arguments other than drop are discouraged" warning
+    suppressWarnings(
+        do.call('[', c(list(x = arr, drop = drop), arglist))
+    )
+}
+
+subsetArrayDimension <- function(dim, idx){
+    dim$any.or.none <- dim$any.or.none[idx]
+    dim$missing <- dim$missing[idx]
+    dim$name <- dim$name[idx]
+    return(dim)
+}
+
+#' Modify cube missing behavior
+#'
+#' By default, CrunchCubes do not show entries for missing categories. You can
+#' include missing values in a `cube` with `showMissing(cube)` and hide them
+#' again with `hideMissing(cube)`.
+#'
+#' @param cube a CrunchCube
+#' @name cube-missingness
+#' @aliases showMissing hideMissing showIfAny
+NULL
+
+#' @rdname cube-missingness
+#' @export
+setMethod("showMissing", "CrunchCube", function(cube) setCubeNA(cube, "always"))
+
+#' @rdname cube-missingness
+#' @export
+setMethod("hideMissing", "CrunchCube", function(cube) setCubeNA(cube, "no"))
+
+#' @rdname cube-missingness
+#' @export
+setMethod("showIfAny", "CrunchCube", function(cube) setCubeNA(cube, "ifany"))
+
+setCubeNA <- function(cube, value = c("always", "no", "ifany")){
+    match.arg(value)
```

Comment:
This should be `value <- match.arg(value)`, pedantically, though it doesn't matter in this case since you control all functions that call this one.